pydmd/pidmd_utils.py

```diff
@@ -0,0 +1,327 @@
+arXiv:2112.04307.
+"""
+import numpy as np
+from numpy.fft import fft, ifft, fft2
+from scipy import sparse
+from scipy.linalg import block_diag, rq
+
+from .utils import compute_svd
+from .rdmd import compute_rank
+
+def compute_unitary(X, Y, svd_rank):
+    """
+    Given the data matrices X and Y and the rank truncation svd_rank, solves
+    for the best-fit unitary operator A that solves the relationship Y = AX.
+    Returns a dictionary containing the corresponding reduced operator atilde.
+    """
+    Ux = compute_svd(X, svd_rank)[0]
+    Yproj = Ux.conj().T.dot(Y)
+    Xproj = Ux.conj().T.dot(X)
+    Uyx, _, Vyx = compute_svd(Yproj.dot(Xproj.conj().T), -1)
+    atilde = Uyx.dot(Vyx.conj().T)
+    return {"atilde": atilde}
+
+def compute_uppertriangular(X, Y):
+    """
+    Given the data matrices X and Y, solves for the best-fit
+    uppertriangular matrix A that solves the relationship Y = AX.
+    Returns a dictionary containing A.
+    """
+    R, Q = rq(X, mode="economic")
+    Ut = np.triu(Y.dot(Q.conj().T))
+    A = np.linalg.lstsq(R.T, Ut.T, rcond=None)[0].T
+    return {"A": A}
+
+def compute_diagonal(X, Y, svd_rank, manifold_opt, compute_A):
+    """
+    Given the data matrices X and Y and the rank truncation svd_rank, solves
+    for the best-fit matrix A that solves the relationship Y = AX and has
+    diagonal entries specified by manifold_opt. Only the eigenvalues and
+    eigenvectors of A are computed if the compute_A flag is not True.
+    Returns a dictionary of computed values.
+    """
+    # Specify the index matrix for the diagonals of A.
+    nx = len(X)
+    if manifold_opt is None:
+        ind_mat = np.ones((nx, 2), dtype=int)
+    elif isinstance(manifold_opt, int) and manifold_opt > 0:
+        ind_mat = manifold_opt * np.ones((nx, 2), dtype=int)
+    elif (isinstance(manifold_opt, tuple)
+            and len(manifold_opt) == 2
+            and np.all(np.array(manifold_opt) > 0)):
+        ind_mat = np.ones((nx, 2))
+        ind_mat[:, 0] *= manifold_opt[0]
+        ind_mat[:, 1] *= manifold_opt[1]
+    elif (isinstance(manifold_opt, np.ndarray)
+            and manifold_opt.shape == (nx, 2)
+            and np.all(manifold_opt > 0)):
+        ind_mat = manifold_opt
+    else:
+        raise ValueError("manifold_opt is not in an allowable format.")
+
+    # Keep track of info for building A as a sparse coordinate matrix.
+    I = []
+    J = []
+    R = []
+
+    # Solve min||Cx-b|| along each row.
+    nxs = np.arange(nx)
+    l1s = (nxs - ind_mat[:, 0] + 1).clip(min=0).astype(int)
+    l2s = (nxs + ind_mat[:, 1]).clip(max=nx).astype(int)
+    for j in range(nx):
+        l1, l2 = l1s[j], l2s[j]
+        C = X[l1:l2].T
+        b = Y[j].T
+        I.append(j * np.ones(l2 - l1))
+        J.append(np.arange(l1, l2))
+        R.append(np.linalg.lstsq(C, b, rcond=None)[0].T)
+
+    # Build A as a sparse matrix.
+    A_sparse = sparse.coo_array(
```

Comment:
`coo_matrix` should be practically equivalent if you just use `A_sparse` to compute the eigenvalues:
```suggestion
    A_sparse = sparse.coo_matrix(
```

Reply:
If I am not mistaken `sparse.coo_matrix` does not have the same constraints as `sparse.coo_array` for the `scipy` version. In that case we can remove such a constraint, otherwise we will abandon py3.7, which will be done anyway, eventually.